esp/services/ws_decoupledlogging/ws_decoupledloggingservice.cpp

```diff
@@ -89,64 +82,122 @@ void CWSDecoupledLogEx::init(IPropertyTree* cfg, const char* process, const char
-            IUpdateLogThread* logThread = createUpdateLogThread(&loggingAgentTree, service, agentName, tankFileDir, loggingAgent);
+            Owned<IUpdateLogThread> logThread = createUpdateLogThread(&loggingAgentTree, service, agentName, tankFileDir, loggingAgent);
             if(!logThread)
                 throw MakeStringException(-1, "Failed to create update log thread for %s", agentName);
-    
+
             CLogRequestReader* logRequestReader = logThread->getLogRequestReader();
             if (!logRequestReader)
                 throw MakeStringException(-1, "CLogRequestReader not found for %s.", agentName);
-            group->loggingAgentThreads.push_back(logThread);
+
+            if (group->getLoggingAgentThread(agentName))
+                throw MakeStringException(-1, "%s: >1 logging agents are named as %s.", groupName, agentName);
+            group->addLoggingAgentThread(agentName, logThread);
         }
-        logGroups.push_back(group);
+        logGroups.insert({groupName, group});
     }
 }
 
 bool CWSDecoupledLogEx::onGetLogAgentSetting(IEspContext& context, IEspGetLogAgentSettingRequest& req, IEspGetLogAgentSettingResponse& resp)
 {
     try
     {
-        BoolHash logAgentFound;
-        StringBuffer errorStatus;
-        StringArray& agentNames = req.getAgentNames();
-        IArrayOf<IEspLogAgentSetting> logAgentSettings;
-        for (auto gp : logGroups)
+        IArrayOf<IEspLogAgentGroupSetting> groupSettingResp;
+        IArrayOf<IConstLogAgentGroup>& groups = req.getGroups();
+        if (!groups.ordinality())
         {
-            const char* tankFileDir = gp->tankFileDir.get();
-            const char* tankFileMask = gp->tankFileMask.get();
-            for (auto in : gp->loggingAgentThreads)
+            for (auto ml : logGroups)
             {
-                IEspLogAgent* agent = in->getLogAgent();
-                const char* agentName = agent->getName();
-                if (!checkName(agentName, agentNames, true))
-                    continue;
+                IArrayOf<IEspLogAgentSetting> agentSettingResp;
+                getSettingsForAllLoggingAgentsInGroup(ml.second, agentSettingResp);
+
+                Owned<IEspLogAgentGroupSetting> groupSetting = createLogAgentGroupSetting();
+                groupSetting->setGroupName(ml.first.c_str());
+                groupSetting->setGroupStatus("Found");
+                const char* tankFileDir = ml.second->getTankFileDir();
+                const char* tankFileMask = ml.second->getTankFileMask();
+                groupSetting->setTankFileDir(tankFileDir);
+                if (!isEmptyString(tankFileMask))
+                    groupSetting->setTankFileMask(tankFileMask);
+                groupSetting->setLogAgentSetting(agentSettingResp);
+                groupSettingResp.append(*groupSetting.getClear());
+            }
+            resp.setSettings(groupSettingResp);
+            return true;
+        }
+
+        ForEachItemIn(i, groups)
```

Comment:
Why is this not in an else, which would enable a single occurrence of resp.setSettings and return? It seems odd to allow the if block to return from within the try block, but the de-facto else block exits the try block before returning.